src/custom_token.rs

```diff
@@ -0,0 +1,534 @@
+
+        impl $crate::export::PartialEq for $ident {
+            fn eq(&self, _other: &Self) -> $crate::export::bool {
+                true
+            }
+        }
+
+        impl $crate::export::Hash for $ident {
+            fn hash<__H: $crate::export::Hasher>(&self, _state: &mut __H) {}
+        }
+    };
+}
+
+// Not public API.
+#[cfg(not(feature = "extra-traits"))]
+#[doc(hidden)]
+#[macro_export]
+macro_rules! impl_extra_traits_for_custom_keyword {
+    ($ident:ident) => {};
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+/// Define a type that supports parsing and printing a multi-character symbol
+/// as if it were a token.
+///
+/// # Usage
+///
+/// As a convention, it is recommended that this macro be invoked within a
+/// module called `punct` or `punctuation` and that the resulting parser be invoked
+/// with a `punct::` or `punctuation::` prefix.
+///
+/// ```
+/// # #[macro_use]
+/// # extern crate syn;
+/// #
+/// mod punct {
+///     custom_punctuation!(LeftRightArrow, <=>);
+/// }
+/// #
+/// # fn main() {}
+/// ```
+///
+/// The generated syntax tree node supports the following operations just like
+/// any built-in punctuation token.
+///
+/// - [Peeking] — `input.peek(punct::LeftRightArrow)`
+///
+/// - [Parsing] — `input.parse::<punct::LeftRightArrow>()?`
+///
+/// - [Printing] — `quote!( ... #path_separator_token ... )`
+///
+/// - Construction from (a) [`Span`]\(s) — `let path_separator_token = punct::LeftRightArrow(sp)`
+///
+/// - Field access to its spans — `let spans = path_separator_token.spans`
+///
+/// [Peeking]: parse/struct.ParseBuffer.html#method.peek
+/// [Parsing]: parse/struct.ParseBuffer.html#method.parse
+/// [Printing]: https://docs.rs/quote/0.6/quote/trait.ToTokens.html
+/// [`Span`]: struct.Span.html
+///
+/// # Example
+// TODO: actual example
+#[macro_export(local_inner_macros)]
+macro_rules! custom_punctuation {
+    ($ident:ident, $($tt:tt)*) => {
+        pub struct $ident {
+            pub spans: [$crate::export::Span; punct_len!($($tt)*)],
+        }
+
+        #[doc(hidden)]
+        #[allow(non_snake_case)]
+        pub fn $ident<__S: $crate::export::IntoSpans<[$crate::export::Span; punct_len!($($tt)*)]>>(
+            spans: __S,
+        ) -> $ident {
+            $ident {
+                spans: $crate::export::IntoSpans::into_spans(spans)
+            }
+        }
+
+        impl $crate::export::Default for $ident {
+            fn default() -> Self {
+                $ident($crate::export::Span::call_site())
+            }
+        }
+
+        impl_parsing_for_custom_punctuation!($ident, $($tt)*);
+        impl_printing_for_custom_punctuation!($ident, $($tt)*);
+        impl_clone_for_custom_punctuation!($ident, $($tt)*);
+        impl_extra_traits_for_custom_punctuation!($ident, $($tt)*);
+    };
+}
+
+// Not public API.
+#[cfg(feature = "parsing")]
+#[doc(hidden)]
+#[macro_export]
+macro_rules! impl_parsing_for_custom_punctuation {
```

Comment:
Some of these new macros will require `#[macro_export(local_inner_macros)]` in order to support 2018 edition. Please try `custom_punctuation!` locally in a 2018 crate and confirm that it works.